docs/design-documents/tools/cmake_components.md

```diff
@@ -0,0 +1,130 @@
+- cmsis
+- events
+- rtos-api
+- drivers
+- hal
+- targets
+- platform
+
+The rest of components and features could be static or object libraries. They could be selected by a user and be added on request.
+
+## Components as static libraries
+
+It is known problem with static libraries - weakly linked symbols are not resolved with strong symbols as someone would have expected. We won't disallow linking with static libraries, just internal Mbed OS components should be object libraries to avoid the problem with weakly symbols as many of the components rely on them.
+
+### Using whole-archive to workaround weak symbols limitation
+
+Toolchains provide a flag to enforce keeping symbols (GCC --whole-archive, clang -force_load). See the issue with forcing linker flag to the components https://github.com/zephyrproject-rtos/zephyr/issues/8441 and https://github.com/zephyrproject-rtos/zephyr/issues/6961 for details. This has drawbacks that will need more research. See also CMake issue https://gitlab.kitware.com/cmake/cmake/-/issues/20078 - not simple to fix in CMake.
+
+`whole-archive` is only available to Gcc Arm, ARMClang does not provide it.
+
+### Removing weak symbols
+
+We could find alternative to weak symbols and fix them in our tree one by one. It was already achieved in Mbed OS 3 where we did not support weak symbols. 
+
+### Object files for files with weak symbols
+
+Mbed 2 was released as a library, we provided object files for files that were providing weak symbols (retarget, file handling and others). They were linked together with the mbed library. This however might not be achievable (its implementation in CMake might contain hacks to get other components paths and flags as retarget file depends on multiple other components).
+
+## External components
+
+They can be either object or static library. Only one limitation due to selecting object library is that any component linked by an application shall not have circular dependencies between the components (CMake will issue an error that we are linking with an object library (mbed-s) and it does not support it - this will be fixed but no ETA given).
+
+## Solution: Mbed-os components
+
+mbed-os component (object library) consists of:
+- cmsis
+- events
+- rtos-api
+- drivers
+- hal
+- targets
+- platform
+
+The rest of the tree is formed by components. 
+
+Each component creates a new CMake library and adds includes/sources:
+
+```
+add_library(mbed-os-nanostack OBJECT)
+
+target_include_directories(mbed-os-nanostack ...)
+
+target_sources(mbed-os-nanostack ...)
+```
+
+If there are dependencies for a component, use `target_link_libraries`. For instance, nanostack depends on 4 other components:
+
+```
+target_link_libraries(mbed-os-nanostack mbed-os-nanostack-libservice mbed-os-netsocket mbed-os-coap mbed-os)
+```
+
+An application just links to what is required:
+
+```
+target_link_libraries(mbed-os-example-nanostack-example mbed-os-nanostack mbed-os)
```

Comment:
do we need to link against mbed-os from an app? if yes, shall mbed-os become mbed-core rather so then mbed-ble mbed-core ?

Reply:
https://gitlab.kitware.com/cmake/cmake/-/issues/18090 - object libraries do not propagate to the final target :/  mbed-os-nanostack is not enough,

Reply:
I'll rename the mbed-os target  to the mbed-os-core. As mbed-os is confusing, it used to be "all in" but with our new approach, it's not true anymore and having `mbed-os` + `mbed-os-component` might confuse some. `mbed-os-core`  it is.